sdk/core/azure-core/inc/azure/core/logger.hpp

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief This header defines the types and functions your application uses to be notified of Azure
+ * SDK client library log messages.
+ */
+
+#pragma once
+
+#include <functional>
+#include <string>
+
+namespace Azure { namespace Core {
+  /**
+   * @brief Log message handler.
+   */
+  class Logger {
```

Comment:
Something went wack y here. Why do we have both Log and Logger classes? We should NOT have these 2 things - this is so confusing!
I thought our plan was to have a Log class with private methods (ShouldWrite & Write) for our internal stuff and public methods for customer-callable stuff (SetLevel & SetListener). Then, an internal header file that made the 2 internal functions (ShouldWrite & Write) callable using just global functions that were freinds of the class.

This seems to have gotten overly complex.

Reply:
Log class is for internal functions - `ShouldWrite()` and `Write()`, the same way as Logger is for `SetListener()` and `SetLevel()`.

Our team really dislikes friend class declarations. This is why there are no private methods in the `Logger` class.

[Other side effect, if we have `Write()` and `ShouldWrite()` as free functions in Internal namespace, `Level` cannot be declared inside `Logger` class, it has to be outside. That's because to declare a function from another namespace, it has to be forward-declared before the class declaration. And in order to even forwarddeclare `ShouldLog()`, `Logger::Level` needs to be known at the point of forward-declaration (because `ShouldLog()` takes it as argument), which is not possible - circular dependecny. But it is always possible to forward-declare `Internal::Log` class, and make it a friend.]

Reply:
@JeffreyRichter, internal functions look better when inside a static class, especially that now it has data members and typedefs ().

I don't think that names are that confusing, given that they do live in different namespaces, and SDK code does not need to interact with the `Logger` class - only with the `Internal::Log` class.

`(Internal::)Log::ShouldWrite()` and `(Internal::)Log::Write()` looks readable to me.
`Logger::SetListener()`, `Logger::SetLevel()`, `Logger::Level`, `Logger::Level::Error` - readable as well.
We write to log, users listen to loggers.